### ISSUE TITLE
Release branching

### ISSUE THREAD

**Issue by ucbjrl** [ucb-bar/chisel-release#14]
As we are about to release a new major version of Chisel/FIRRTL (3.2.0, 1.2.0), it is time to reconsider our current branch naming policy. We currently support a single, major development branch (`master`) per project, and multiple older version development branches (`3.0.x`, `3.1.x`). We currently have a single release branch (`release`) per project, corresponding to the current major release version (typically, one major version behind `master`). There may be separate release branches for older releases, but (at least in principle), these could be recreated by branching from `release` at the appropriate tag. Current documentation suggests making releases from the appropriate version development branch (`master` or `3.1.x`), and updating the versions in `build.sbt` appropriately. Should we change this to permit multiple release branches, or incorporate a more complex flow into our release process?

Experience indicates it is convenient to leave the versions in the `3.1.x` branch as SNAPSHOTs for that version. This facilitates testing prior to release as newer versions of sbt will not overwrite existing non-SNAPSHOT versions during publish (either `publishLocal` or `publishSigned`). This in turn suggests that prior to a release, the appropriate development version is merged into the release branch, and the versions in `build.sbt` updated to reflect the actual version numbers for this release. Since we are about to support multiple versions, it may be time to name version-specific release branches. In general, branch `3.1.release` would be the branch for 3.1 releases (with each release tagged appropriately), and branch `3.1.x` would be its development branch. Typically, once the first API breaking change is added to `master`, its SNAPSHOT version is bumped, and its state (prior to the first API breaking change) becomes the head of the prior version development branch, in this case, `3.2.x`.

We currently limit development in prior versions to bug fixes or future compatibility enhancements so there is typically little development in the `3.1.x` branch. But, branching is cheap, and the overhead incurred by keeping a "hot fix" branch warmed up doesn't appear to be too onerous. One could argue that we don't really need both `3.1.release` and `3.1.x` branches since one can be created from the other by changing the versions in `build.sbt`. I think it does simplify the tooling to keep these branches distinct, but it could be handled by keeping the actual release versions in a separate file (under version control, updated at release) and a script to switch between the development and release version by editing `build.sbt` appropriately.

Alternatively, (with a minor change to our development versions) we could use the [sbt release plugin](https://github.com/sbt/sbt-release) which:
 - sets the release version (drops the "-SNAPSHOT" from the version)
 - commits the release version
 - tags the release
 - publishes artifacts
 - updates the version (restoring the trailing "-SNAPSHOT")
 - commits the next version
 - pushes changes

I've been a little wary of using this since:
 - it does an awful lot and although it claims the release process is aborted if any intermediate task fails, I've never had the confidence to test it,
 - I am a little wary of keeping release and development code in the same branch, although it looks like there is sufficient control to spin off a hot fix branch if required,
 - it's not just the version of the published project that needs to be managed, but also its dependencies

Following this route would make it easier for others to generate releases, but does tend to make us more dependent on sbt. Of course, we could wrap up this process (and tailor it for our current versioning) in the `chisel-release/Makefile`.

The solution to this issue is:
- maintain separate `z.y.release` and `z.y.x` branches the heads of which are identical except for versions in `build.sbt`, or
- maintain a single `z.y.x` branch which contains both development and release code with releases tagged at the version containing `build.sbt` with explicit, non-SNAPSHOT versions, and a more sophisticated release process that mimics the behavior of `sbt release`.

**Comment by chick** [comment]
This all looks pretty reasonable to me. My main question would be: how does this scale across Big-N number of projects. I guess the dependency graph is described in the sbt files, but does that make it easy enough to reason about. Maybe this is no worse in that regard than our current practices but I am not that clear about what it will take to generate build environments of users reporting problems

**Comment by ucbjrl** [comment]
Good question. We'll need their corresponding build.sbt (or build.sc, or the configuration file for whatever build tool they're using), or the paths to the versioned JARs, which determine the version of the projects incorporated in their build. These versions should correspond to tags in the associated repositories (regardless of whether there are multiple release branches or not). If you checkout the code corresponding to the tag in each repository, you should have the same BIGN code they're working with.
Basically, you should be able to treat each repository independently.
Once you have a fix, you'll want to publish new releases of each affected repository (bumping their versions if they're stable releases), and new versions of every downstream repository so that they depend on the new (fixed) version of the affected repositories.
